### PR TITLE
Fixing a memory leak in map mode.

### PR DESCRIPTION
**Pull request by authenticate**
Hi,

This is a small fix for a memory leak in map mode.

The following memory allocations were not cleaned up:

```
for (uint8_t i = 1; i < enemy_number; ++i) {
         EnemySprite* copy = new EnemySprite(*enemy);
         copy->Reset();
         _enemies.push_back(copy);
}
```

This happens in the "EnemyZone::AddEnemy" function when extra copies of the enemy are requested by the second parameter.  Usually, this happens from the script files.

I fixed it by adding an extra vector to keep track of these extra allocations made by the zone class.  I do not particularly like this solution.  I think there is a better way to formally solve it.  However, in the short term, it at least fixes the memory leak.

Finally, I made some minor "override" keyword and "explicit" keyword changes to sections of the code base I searched through while trying to diagnose the leak.

Let me know if you have any questions or any issues come up.

Thanks.

### CODE REVIEW

File: src/modes/map/map_zones.cpp [hunk 1]
```diff
@@ -257,10 +297,11 @@ void EnemyZone::AddEnemy(EnemySprite* enemy, uint8_t enemy_number)
     _enemies.push_back(enemy);
 
     // Create any additional copies of the enemy and add them as well
-    for(uint8_t i = 1; i < enemy_number; ++i) {
+    for (uint8_t i = 1; i < enemy_number; ++i) {
```
Comment (authenticate) [inline src/modes/map/map_zones.cpp]: This is where the leak occurred.

File: src/modes/map/map_zones.cpp [hunk 2]
```diff
@@ -238,6 +262,22 @@ EnemyZone::EnemyZone(uint16_t left_col, uint16_t right_col,
...
 }
 
+EnemyZone::~EnemyZone()
+{
+    if (_spawn_zone != nullptr) {
+        delete _spawn_zone;
+        _spawn_zone = nullptr;
+    }
+
+    for (auto& enemy_owned : _enemies_owned) {
```
Comment (authenticate) [inline src/modes/map/map_zones.cpp]: This is a quick fix to clean up the memory.

**Comment by Bertram25** [comment]
I agree with you, this is a design flaw. I'll add an issue not to forget about it. I'll test the code now.

**Comment by Bertram25** [comment]
Works fine, I'll merge. Thanks! :)

**Pull request merged**